fix(synthetics): export monitor ID in all synthetic monitor resources

Aashirwadjain opened the pull request:
# Description

This PR aims to add support to Synthetic Monitor Resources to export monitor id.
Github Issue: https://github.com/newrelic/terraform-provider-newrelic/issues/2814

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

## Checklist:

Please delete options that are not relevant.

- [x] My commit message follows [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/)
- [x] My code is formatted to [Go standards](https://go.dev/blog/gofmt)
- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes. Go [here](https://github.com/newrelic/terraform-provider-newrelic/blob/main/CONTRIBUTING.md#testing) for instructions on running tests locally.

pranav-new-relic commented:
@Aashirwadjain looks like we have some conflicts. You might want to raise a PR from a branch updated with the latest on main.

<img width="998" alt="image" src="https://github.com/user-attachments/assets/3fc30f9d-ea01-4244-85fa-967081d216d9" />

vagrawal-newrelic commented:
@Aashirwadjain Please update the docs also to reflect that these resources are exporting `monitor_id` also.
Provided the part of the doc where you need to add `monitor_id` argument

[synthetics_broken_links_monitor.html.markdown](https://github.com/newrelic/terraform-provider-newrelic/blob/add-synthetic-monitor-id-support/website/docs/r/synthetics_broken_links_monitor.html.markdown#attributes-reference)

[synthetics_cert_check_monitor.html.markdown](https://github.com/newrelic/terraform-provider-newrelic/blob/add-synthetic-monitor-id-support/website/docs/r/synthetics_cert_check_monitor.html.markdown#attributes-reference)

[synthetics_monitor.html.markdown](https://github.com/newrelic/terraform-provider-newrelic/blob/add-synthetic-monitor-id-support/website/docs/r/synthetics_monitor.html.markdown#attributes-reference)

[synthetics_script_monitor.html.markdown](https://github.com/newrelic/terraform-provider-newrelic/blob/add-synthetic-monitor-id-support/website/docs/r/synthetics_script_monitor.html.markdown#attributes-reference)

[synthetics_step_monitor.html.markdown](https://github.com/newrelic/terraform-provider-newrelic/blob/add-synthetic-monitor-id-support/website/docs/r/synthetics_step_monitor.html.markdown#attributes-reference)

Aashirwadjain commented:
@vagrawal-newrelic Added the `monitor_id` attribute in Attributes Reference section of the documents that you mentioned.

pranav-new-relic commented:
@Aashirwadjain all good so far, minor suggestions I'd like you to implement -

- It's important to cover whatever we can with tests, and now that `monitor_id` is a new computed attribute, it'd be a good idea to include checks for this in our integration tests to make sure the `monitor_id` is set during create and matches the `monitor_id` fetched during read; which kind of eliminates the onus on us to test this functionality on all seven kinds of Synthetic monitors, as the test case will make sure the `monitor_id` is stable across create, read and update 👍 
   - the hint to do this: for each Synthetic monitor type, identify the corresponding integration test file in the code, in which you'll find a bunch of test cases corresponding to that monitor, and a common check function that is consumed by all of these test cases - which is where you'd want to make this change to add a validation check as specified above. You can refer to the check that's already in place with GUID, and you'll need to do something similar with `monitor_id`. This should not be a complex change - should not go over 3-4+ lines so it shouldn't take you very long to do this.

- In the documentation changes made to add `monitor_id` as an exported attribute, just add a phrase before the `.` `"(not to be confused with the GUID of the monitor)"` - of course, this is pretty clear as we have the mention of `(GUID)` against `id`, but this is just to make it extra clear.

Once the changes in (1) are done, all test cases should be passing here; that should be the signal for us to go ahead with this PR - let me know once you see this and I'll approve ✅

pranav-new-relic commented:
@Aashirwadjain while the changes in https://github.com/newrelic/terraform-provider-newrelic/pull/2854/commits/0c8e81afb16f2b28d28a3f2bc51de916b8b3b832, this is not an optimal solution. Fetching tags is an unnecessary expense, additionally given that `SyntheticMonitorEntity` in the Go Client has the attribute `MonitorId` by default. In a structure similar to wherever we're matching GUID in these functions, just add an update under them to check for MonitorId and that should be enough. We don't need to fetch all tags and iterate over them for something that's already readily available as an attribute

Also please sync your branch with main

pranav-new-relic commented:
Also, in your next commit, please add the latest version of the Go Client to push @rpaliwal1997's changes

Aashirwadjain commented:
Hi @pranav-new-relic, Thanks for the suggestions. I have included the `monitor_id` attribute in the test cases for synthetic monitor resources as you mentioned. Also I have updated the approach of tags, the validation in test cases are performed by fetching `monitor_id` directly from `SyntheticMonitorEntity`. Apart from this, I have also updated the Go client version to `v2.56.1`.

vagrawal-newrelic reviewed (approved):
LGTM
